Feature/mbqc

NewGitter2017 opened the pull request:
Just a first version of MBQC

NewGitter2017 committed:
Adding MBQC support

Adding a preliminary version of the MBQC conversion module.

NewGitter2017 committed:
Tracking inputs/outputs

Instantiated the class and added a map which keeps track of the current label of the input/output qubits.

NewGitter2017 committed:
Added tests

NewGitter2017 committed:
Fixed mapping for 'split_subgraphs()'

Fixed mapping for 'split_subgraphs()' and added a test for it.

NewGitter2017 committed:
Finished the single_conversion method - needs testing

NewGitter2017 committed:
Checked that the single_conversion() method works

Did some bugfixes on single_conversion() and checked that it does what intended.

NewGitter2017 committed:
Fixed a maths mistake

I was adding a Hadamard to the outputs which shouldn't be there.

NewGitter2017 committed:
Improved tracking of i/o maps and added more tests

Ready for code review

sjdilkes commented on pytket/pytket/mbqc_compile/__init__.py:
```diff
@@ -0,0 +1,41 @@
+"""Functions for splitting circuits to a variable number of subcircuits and 
+converting each to measurement based quantum computing."""
+
+#import numpy as np
```
Please can you remove all these imports

NewGitter2017 replied:
Done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+
+from pytket.transform import Transform
+from pytket.circuit import Circuit as Circuit
+#from pytket.pyzx import pyzx_to_tk
```
remove unused import

NewGitter2017 replied:
Done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+from pyzx.graph.graph_s import GraphS as Graph
+from pyzx.circuit.graphparser import circuit_to_graph
+from pyzx.gflow import gflow as gflow
+#from typing import None
```
remove unused import

NewGitter2017 replied:
Done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+    """
+    def __init__(self, c: Circuit) -> None:
+        """
+        :param c:       A pytket circuit.
```
Can you change this to be more descriptive (as the type declaration already covers what it is).

NewGitter2017 replied:
Done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        self.inputs = {}
+        self.outputs = {}
+        for qubit in c.qubits:
+            self.inputs[qubit.index[0]] = qubit.index[0]
```
Is it possible to index the keys & values of this dictionary with the actual Qubit type instead of the indexing? We can't guarantee that the values returned by qubit.index[0] will all be unique (different registers may have qubits labelled differently but with the same index).

NewGitter2017 replied:
Done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        for qubit in c.qubits:
+            self.inputs[qubit.index[0]] = qubit.index[0]
+            self.outputs[qubit.index[0]] = qubit.index[0]
+        pass
```
remove `pass`

NewGitter2017 replied:
Can't find it in the refactor version, I'm inclined to assume that I got rid of it at some point

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+    
+    def single_conversion(self) -> Circuit:
+        """
+        Converts a tket circuit to another with reduced depth and higher width.
```
tket -> pytket

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        c2.add_circuit(c3,c3.qubits,c3.bits[:len(c3.qubits)])
+        return c3
+    
+    def zx_convert(c: Circuit) -> pyzxCircuit:
```
Can you declare this method as static (`@staticmethod`)

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        :returns:       A pytket circuit.
+        :rtype:         Circuit
+        """
+        pyzxc = MPattern.zx_convert(self.c)
```
Why pass an MPattern attribute (self.c) to a static MPattern method? Why not make `zx_convert` non-static, and modify the circuit held at self.c?

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+from pytket.extensions.pyzx import tk_to_pyzx
+from pyzx.simplify import interior_clifford_simp
+from pyzx.circuit import Circuit as pyzxCircuit
+from pyzx.graph.graph_s import GraphS as Graph
```
why import as Graph instead of GraphS? Maybe counterintuitive, but there is a Graph class in pytket already so better to keep the "S" for easier distinction.

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+                    self.outputs[o] = v
+                    break
+    
+    def entangle(g: Graph) -> Circuit:
```
`@staticmethod`

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        #At this point we have the new qubit register, the CZ gates, and the
+        #mapping for each input/output but we are missing the conditional
+        #measurements.
+        subs = self.split_subgraphs(g)
```
why run `split_subgraphs` after `entangle`?

NewGitter2017 replied:
"split_subgraphs" is needed for the measurement&corrections circuit while "entangle" is needed for the CZ circuit

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        :returns:       A circuit containing measurements and conditional gates.
+        :rtype:         Circuit
+        """
+        S = {}
```
Please can you give this object a more descriptive name

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        """
+        pyzxc = MPattern.zx_convert(self.c)
+        g = self.zx_diagram(pyzxc)
+        c2 = MPattern.entangle(g)
```
Please give c2 (and c3) more descriptive names (cz_entangling_circuit, measurement_correction_circuit, etc)

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        :param type:    Circuit
+        """
+        self.c = c
+        self.qubits = len(c.qubits)
```
similarly to comment below, why I understand this is used to declare a range later, why can't this be `List[Qubit]`? And Qubit objects used to index things?

NewGitter2017 replied:
done.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        """
+        g = circuit_to_graph(pyzxc)
+        interior_clifford_simp(g, quiet=True)
+        for q in range(self.qubits):
```
replace with iterating over Qubit object instead of indices?

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        g = circuit_to_graph(pyzxc)
+        interior_clifford_simp(g, quiet=True)
+        for q in range(self.qubits):
+            self.inputs[q] = sorted(list(g.vertices()))[q]
```
Using Qubit instead of indices would mean this would be need to be updated. But it will be a safer solution long run.

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        self.label_squish(g)
+        return g.copy()
+    
+    def label_squish(self, g: Graph) -> None:
```
similarly amend to use Qubit instead of indices

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        :returns:       A pytket circuit.
+        :rtype:         Circuit
+        """
+        c = Circuit(len(g.vertices()),len(g.vertices()))
```
If using Qubit instead of indices, this will need to initialise by itearting through all qubits and adding them to the Circuit

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        vlist.reverse()
+        edge_pool = set(g.edge_set())
+        finished_edge_pool = {}
+        doneround = {}
```
could you add a comment to describe this name?

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        finished_edge_pool = {}
+        doneround = {}
+        for v in vlist:
+            doneround[v] = False
```
doneround = {v: False for v in vlist}

sjdilkes replied:
Alternatively, instead of using a dict bool values, you could use a set and add elements to it once they've been processed

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        doneround = {}
+        for v in vlist:
+            doneround[v] = False
+        while len(edge_pool)>0:
```
Could you document these loops please? Just in line comments

sjdilkes replied:
Also, would it make more sense for this whole process to loop through Edges in edge_pool front to back? `for edge in edge_pool` do x?

NewGitter2017 replied:
Not sure, this method was designed to produce minimum depth CZ circuits by prioritizing vertices with more edges at any given timestep so the loops might look a bit weird... Not sure how exactly it works at this point

sjdilkes replied:
ah, so the iterating here is crucial to results? in which case can ya make that clear - maybe even make a separate method

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+            doneround[v] = False
+        while len(edge_pool)>0:
+            for vid in range(len(vlist)):
+                if not doneround[vlist[vid]]:
```
why vlist[vid]?

NewGitter2017 replied:
I believe we discussed this about 2 months ago but we might need to revisit after the latest review.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+            for vid in range(len(vlist)):
+                if not doneround[vlist[vid]]:
+                    for vid2 in range(vid+1,len(vlist)):
+                        if ((not doneround[vlist[vid]]) and (not doneround[vlist[vid2]])):
```
do we need to check `doneround[vlist[vid]` again?

NewGitter2017 replied:
Fixed.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+            c.add_barrier(range(len(g.vertices())), range(len(g.vertices())))
+        return c
+
+    def remove_redundant(self, g: Graph) -> None:
```
Is this not something already supported by pyzx?

NewGitter2017 replied:
We discussed this a while back, I'm basically merging some of the boundary vertices connected with simple edges which pyzx didn't get rid of.

sjdilkes replied:
ha yeah I wrote this on 19 Jan

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,434 @@
+        self.identity_cleanup(g)
+        pass
+
+    def identity_cleanup(self, g: Graph) -> None:
```
Similarly does this kind of cleanup just not exist in pyzx?

NewGitter2017 replied:
Basically using remove_redundant() might open the door to some further simplifications which identity_cleanup() takes care of. I can't pass it back to pyzx because remove_redundant() changes the format of the graphs slightly.

sjdilkes commented on pytket/tests/mpattern_test.py:
```diff
@@ -0,0 +1,114 @@
+from pytket.circuit import Circuit as Circuit
+from pytket.mbqc_compile.mpattern import MPattern
+#import pytest  # type: ignore
+#import platform
```
remove unused imports

NewGitter2017 replied:
Done.

sjdilkes reviewed (changes requested):
I've made a bunch of first comments. I've left reviewing further for now as I suspect changes willl be made in line with using Qubit objects for indexing vs indices.

NewGitter2017 committed:
Made sure the new mapping works for arbitrary circuit

Reworked mapping, tested it and implemented a lot of changes from the code review

NewGitter2017 committed:
Added prototype of multi_conversion()

Added the first version of multi_conversion() but it needs to be tested and also it just returns a list of tuples, with each tuple containing a circuit and an i/o map dictionary. These have to be merged together into a single new circuit which it doesn't do yet.

NewGitter2017 committed:
Wrote tests for multi_conversions()

Wrote tests, fixed bugs and added comments and descriptions to the new methods.

NewGitter2017 committed:
Routing conversion half-done

NewGitter2017 committed:
Made the MPattern class work with new routing

Reworked, tested and cleaned up MPattern class to work with new routing. Still need to add comments and clarify some of the tests and then proceed to benchmarking.

NewGitter2017 committed:
Added descriptions and tests for all the methods.

Cleaned up the code - Ready to start benchmarking.

NewGitter2017 committed:
Working version

NewGitter2017 committed:
Code cleaned up for review

Cleaned up the mpattern class for review.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+class MPattern:
+    """
+    Class with tools to convert a pytket circuit into a new pytket circuit
+    with lower depth and higher width, by using MBQC techniques.
```
write out "Measurement Based Quantum Computation" here for clarity

NewGitter2017 replied:
Done.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        """
+        Converts a pytket circuit to another with reduced depth and higher width.
+        
+        :returns:       A tuple containing the new circuit and the i/o map.
```
Please write "input output map" and explain what this is

NewGitter2017 replied:
Done.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :returns:       A tuple containing the new circuit and the i/o map.
+        :rtype:         tuple
+        """
+        (g,io_map) = self.zx_diagram() #Creates a simplified ZX diagram.
```
Typically if a class method references another class method I'd expect the referenced method to be defined first in the class - please could you move zx_diagram() and split_subgraphs() definitions above single_conversion() in the class layout (and generally check all defined methods and move them around to respect this).

NewGitter2017 replied:
Done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        (g,io_map) = self.zx_diagram() #Creates a simplified ZX diagram.
+        subs = self.split_subgraphs(g,io_map) #Returns list of disjoint subgraphs.
+        cz_circ = MPattern.entangle(g) #Creates the CZ circuit part of the pattern.
+        m_circ = MPattern.correct(subs) #Circuit implementing measurements/corrections.
```
As fun as it is that correct and entangle are static, you may aswell remove the MPattern here

NewGitter2017 replied:
Done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        subs = self.split_subgraphs(g,io_map) #Returns list of disjoint subgraphs.
+        cz_circ = MPattern.entangle(g) #Creates the CZ circuit part of the pattern.
+        m_circ = MPattern.correct(subs) #Circuit implementing measurements/corrections.
+        cz_circ.add_circuit(m_circ,[])
```
I guess this works with an empty list, though this seems like a potential vulnerability to me, could MPattern.correct return the list of qubits for adding the circuit wires to?

NewGitter2017 replied:
Discussed

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        cz_circ.add_circuit(m_circ,[])
+        return (cz_circ, io_map)
+    
+    def multi_conversion(self, n: int = 1, strategy: str = "Gates", ) -> list:
```
Would be nice to change strategy to take an ENUM instead of a str

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+    def multi_conversion(self, n: int = 1, strategy: str = "Gates", ) -> list:
+        #Currently 'strategy' takes a 'str' type parameter that is either "Depth"
+        #or "Gates". Might want to consider additional strategies and switch to
+        #an enum in the future.
```
ha, yeah this!

NewGitter2017 replied:
Not sure what that is... probably some remnant of something I had removed but it doesn't seem to exist in the refactored version so "Done" I guess?

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :returns:        A list of tuples containing circuits and i/o maps.
+        :rtype:          list
+        """
+        depth_structure = self.depth_structure()
```
similarly to before, move this method ahead of multi_conversion in class structure

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :rtype:          list
+        """
+        depth_structure = self.depth_structure()
+        size = len(depth_structure)
```
size of what? please make variable name slightly more descriptive

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+            non_cliff = 0
+            for d in depth_structure:
+                for gate in d:
+                    if not MPattern.is_Clifford(gate):
```
`MPattern.is_Clifford` -> `is_Clifford`

NewGitter2017 replied:
changed after refactoring

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        """
+        depth_structure = self.depth_structure()
+        size = len(depth_structure)
+        if strategy == "Gates":
```
for now with string add a check that the string is one of "Gates" "Clifford" or "Depth" else throw error. Enum will fix this problem.

NewGitter2017 replied:
changed to enum after refactoring

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        slice_size = math.ceil(size/n)
+        done_depth = 0
+        output = []
+        if strategy == "Depth":
```
Note that you can assign functions to enums which will make this code neater, meaning you'll just have one function call that will call one of these strategies.

NewGitter2017 replied:
done over the refactoring

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        done_depth = 0
+        output = []
+        if strategy == "Depth":
+            for curr in range(n):
```
`curr` -> `_`

NewGitter2017 replied:
This sort of thing probably exists in a bunch of places, should I rename all such loop variables to "_"?

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+                        subcircuit.add_gate(Op=gate.op, args=gate.args)
+                sub_pattern = MPattern(subcircuit)
+                output.append(sub_pattern.single_conversion())
+                if finish_at >= size:
```
change loop to while loop with this condition?

NewGitter2017 replied:
Well, if I remember what was going in my head correctly, the primary loop condition is "for curr in range(splits):"... Basically we want to split it into "splits" segments originally. However, "splits" only detemines the maximum number of segments allowed. The if statement is there as a secondary condition to exit the loop early if we reach the end of the circuit with fewer segments

Basically, let's say you have a depth 10 circuits and you want to split it into 7 segments... Well, 10/7~=1.43 but that is gonna be rounded up to 2... So we will be splitting the circuit into segments of depth 2... But after 5 segments we reach the end of the circuit so that if statement will ensure we don't actually attempt to add more than 5 segments.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+                    subcircuit.add_bit(bit)
+                for depth_list in depth_structure[done_depth:finish_at]:
+                    for gate in depth_list:
+                        subcircuit.add_gate(Op=gate.op, args=gate.args)
```
does add_gate use like this cover rotation gates/symbolic gates or barrier operations?

NewGitter2017 replied:
I think it does. Rotation gates should be included, barriers maybe not but we have removed barriers altogether from the current implementation.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+                added_depths = 0
+                stop_at_next_nClifford = False
+                stopped = False
+                for depth in depth_structure[done_depth:]:
```
maybe `for slice` instead of `for depth`? slightly more descriptive I think

NewGitter2017 replied:
well, I just tried that but it seems that "slice" already means something in python

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        if strategy == "Depth":
+            for curr in range(n):
+                finish_at = min(done_depth + slice_size,size)
+                subcircuit = Circuit()
```
this subcircuit initialisation is identical for each str if detection, please move out to just below output = []

NewGitter2017 replied:
I may have misunderstood what you mean, but shouldn't the subcircuit initialisation stay in the loop? Each iteration of the loop needs to create a new subcircuit object, then add gates to it and then add it to the list of subcircuits.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+                    subcircuit.add_qubit(qubit)
+                for bit in self.c.bits:
+                    subcircuit.add_bit(bit)
+                for depth_list in depth_structure[done_depth:done_depth+added_depths]:
```
rather than working out the added depth in advance and then reiterating through slices to add gates, can it add gates as it goes?
additionally, rather than persisting depth_structure and using indexing to work out slices, could each conversion stage consume part of depth_structure? so you're always working from a zero index

NewGitter2017 replied:
Well, the main problem is that whether a slice gets added to the current subcircuit or not may depend on whether you will find a non-Clifford gate or not, or even on how many non-Clifford gates this slice contains.

This means that as you are looking at the gates, you can't add them directly until you know you are supposed to... That's why it iterates twice, once to figure out how many slices are gonna be added to the current segment and another loop to actually add all the gates in those slices.

As for the second part of your comment, well the loop used here is "for depth in depth_structure[done_depth:]:" which means that for each segment I am already only giving what's left of the depth_structure... Do you mean that at the end of each iteration I should use something like "depth_structure = depth_structure[done_depth]" so as to completely discard the previous bit?

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+                else:
+                    done_depth += added_depths
+        elif strategy == "Clifford":
+            if size > 0:
```
the requirement of this condition implies to me that the logic can be changed

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        elif strategy == "Clifford":
+            if size > 0:
+                clifford_circ = True
+                for gate in depth_structure[0]:
```
should this be depth_structure[done_depth]?

sjdilkes replied:
or should this be checking the whole depth structure?

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+                    output.append(new_tuple)
+        return output
+        
+    def unrouted_conversion(self, n: int = 1, splitStrat: str = "Gates") -> tuple:
```
return type is Tuple[Circuit, Dict[Qubit, Qubit]] (I think, or at least close).  Mypy should help you fix all of these though.

sjdilkes replied:
splitStrat should be ENUM

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        pattern_list = self.multi_conversion(n, splitStrat)
+        new_c = Circuit()
+        prev_map = {}
+        for p in range(len(pattern_list)):
```
this whole method could do with better commenting coverage as its hard to track as is. Please can you add?

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :param splitStrat: Splitting strategy either by "Depth" or by "Gates".
+        :param type:       str
+        
+        :param routeStrat: Routing strategy either by "Separate" or "Sequential".
```
Enums

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :returns:          A tuple containing a circuit and an i/o map dictionary.
+        :rtype:            tuple(Circuit,dict)
+        """
+        if (type(arch)==type(FullyConnected(0))) or (arch == None):
```
We shouldn't set `None` as the default argument here -> if the user passes something that isn't an Architecture that is a user fault. 
Similarly with passing a FullyConnected architecture, though im assuming FullyConnected(0) in particular causes some break - what is this?

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+            elif routeStrat == "Sequential":
+                return self.routed_conversion_sequential(pattern_list,arch)
+        
+    def routed_conversion_separate(self, pattern_list: list, arch: Architecture = None) -> tuple:
```
remove None default

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        onto some given device architecture and then generates a network of SWAP gates
+        which also respect the architecture, to join each segment to the next one.
+        Finally returns a single tuple containing the resulting Circuit object and
+        the dictionary of the final i/o map.
```
i/o -> input to output

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :param arch:         The architecture to route onto.
+        :param type:         Architecture
+        
+        :returns:            A tuple containing a circuit and an i/o map dictionary.
```
i/o -> input to output

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :returns:            A tuple containing a circuit and an i/o map dictionary.
+        :rtype:              tuple(Circuit,dict)
+        """        
+        new_c = Circuit()
```
please can you reannotate this method with comments?

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        new_c = Circuit()
+        for q in arch.nodes:
+            new_c.add_qubit(q)
+        for p in range(len(pattern_list)):
```
maybe change p to `index`?

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        for p in range(len(pattern_list)):
+            (curr_circ,curr_map) = pattern_list[p]
+            route_map = {}
+            if p==0:
```
can work required in the. p==0 case be done outside for for p in range(...) loop? to avoid this if else case

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+                    curr_map["i"][k] = prev_map["o"][k]
+                curr_circ.rename_units(route_map)
+                """
+                output_list = [curr_circ.qubits[q] for q in list(curr_map["o"].values())]
```
I'm guessing this can be deleted

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :returns:        The result of the check.
+        :rtype:          bool
+        """
+        if aGate.op.get_name() in {'Z','X','Y','H','S','V','Sdg','Vdg','SX','SXdg','CX','CY','CZ','CH','CV','CVdg','CSX','CSXdg','CCX','SWAP','CSWAP','noop','BRIDGE','Reset'}:
```
I think you can just use `aGate.op.is_clifford_type()` here

NewGitter2017 replied:
I believe Yao mentioned at some point that this worked for some cases like X, Y, Z gates but not for others like Rz(pi/2), Rx(3pi/2).

The method I wrote isn't by any means exhaustive but I believe at least covers the ones that I was told might not be included which is why I left it in

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        qubits = self.c.qubits
+        depth = self.c.depth()
+        qn = self.c.n_qubits
+        current_frontiers = [0]*qn
```
current_frontier = [0]*len(qubits), remove qn

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        depth = self.c.depth()
+        qn = self.c.n_qubits
+        current_frontiers = [0]*qn
+        depth_slices = [[] for d in range(depth)]
```
range(self.c.depth())

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        qn = self.c.n_qubits
+        current_frontiers = [0]*qn
+        depth_slices = [[] for d in range(depth)]
+        for gate in gates:
```
for gate in self.c.get_commands()

NewGitter2017 replied:
Done.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        :rtype:         list
+        """
+        gates = self.c.get_commands()
+        qubits = self.c.qubits
```
is this used?

NewGitter2017 replied:
Lost track of this code snippet in the refactored version

NewGitter2017 replied:
Not sure why that was there... There doesn't seem to be any import of pyzxCircuit objects in the refactored version at least.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        #pyzx is updated such that graph.copy() no longer changes vertex labels
+        #then comment out the next line (label_squish(g)).
+        self.label_squish(g,io_map) #Squishes the labels of the graph vertices to fill in empty vertex indices.
+        return (g.copy(),io_map)
```
why copy?

NewGitter2017 replied:
Duplicate reply by basically when I'm identifying and splitting disjoint subgraphs the easiest way I found to do it is to go through the original graph once to separate the vertices into X clusters, then create X copies of the original graph and in each copy just remove all the vertices that don't belong in the corresponding cluster

sjdilkes replied:
fair enough -> can you include comments to state this please

NewGitter2017 replied:
done

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+        interior_clifford_simp(g, quiet=True) #Remove interior clifford spiders.
+        new_inputs = {}
+        new_outputs = {}
+        sorted_vertices = sorted(list(g.vertices()))
```
why sort?

NewGitter2017 replied:
They are probably already sorted by index but I didn't want to take the chance.

sjdilkes commented on pytket/pytket/mbqc_compile/mpattern.py:
```diff
@@ -0,0 +1,1074 @@
+    def label_squish(self, g: GraphS, io_map: dict) -> None:
+        """
+        Updates the input/output labels of the MPattern to matched a squished
+        graph caused by g.copy().
```
can we define the behaviour outside of g.copy? and why is g.copy required?

NewGitter2017 replied:
basically when I'm identifying and splitting disjoint subgraphs the easiest way I found to do it is to go through the original graph once to separate the vertices into X clusters, then create X copies of the original graph and in each copy just remove all the vertices that don't belong in the corresponding cluster

sjdilkes replied:
fair enough - can you comment making this clear please

NewGitter2017 replied:
done

EDIT: Added an explanation to the descriptions of split_subgraphs and label_squish
